Update release notes for v2.1.0 (#805)
dotnet/SqlClient: release-notes/2.1/2.1.0.md

# Release Notes

## Microsoft.Data.SqlClient 2.1.0 released 19 November 2020

This update brings the below changes over the previous release:

### Added
- Microsoft.Data.SqlClient symbols are now source-linked [#789](https://github.com/dotnet/SqlClient/pull/789) [[Read more](#Source-Linking-of-Microsoft.Data.SqlClient-symbols)]
- Added new API to clear cached access tokens from token provider [#800](https://github.com/dotnet/SqlClient/pull/800)
- Added implementation of `SqlFacetAttribute` [#757](https://github.com/dotnet/SqlClient/pull/757)

### Fixed
- Fixed `InvalidOperationException` and `NotSupportedException` errors from driver due to `WriteAsync` collisions [#796](https://github.com/dotnet/SqlClient/pull/796)

> - Fixed bug in `ExecuteXmlReaderAsync` implementation [#782](https://github.com/dotnet/SqlClient/pull/782)

Reviewer: ```suggestion
- Fixed incorrect Settings.Async flag in `ExecuteXmlReaderAsync` [#782](https://github.com/dotnet/SqlClient/pull/782)
```

> - Fix a regression in Windows Integrated authentication when using managed networking [#777](https://github.com/dotnet/SqlClient/pull/777)

Reviewer: ```suggestion
- Fixed a regression in Windows Integrated authentication when using managed networking [#777](https://github.com/dotnet/SqlClient/pull/777)
```

- Fixed Bulk Copy Async deadlock issues with custom `IDataReader` when using `SqlDataReader` internally [#779](https://github.com/dotnet/SqlClient/pull/779)

> - Fix serialization issue with `SqlException` in .NET Core [#780](https://github.com/dotnet/SqlClient/pull/780)

Reviewer: ```suggestion
- Fixed a serialization issue with `SqlException` in .NET Core [#780](https://github.com/dotnet/SqlClient/pull/780)
```

### Changes
- Updated versions of `Microsoft.IdentityModel` package dependencies [#794](https://github.com/dotnet/SqlClient/pull/794)

## Summary of changes in 2.1

> All changes in Microsoft.Data.SqlClient v2.1 over v2.0:

Reviewer: ```suggestion
All changes in Microsoft.Data.SqlClient v2.1 since v2.0:
```

### Added
- Added support for Always Encrypted on all supported platforms for .NET Standard 2.0 [#756](https://github.com/dotnet/SqlClient/pull/756) [[Read more](#Cross-Platform-support-for-Always-Encrypted-v1)]
- Added support for Always Encrypted with secure enclaves on Unix for .NET Core 2.1+ and on all supported platforms for .NET Standard 2.1+ [#676](https://github.com/dotnet/SqlClient/pull/676) [[Read more](#Cross-Platform-support-for-Always-Encrypted-with-Secure-Enclaves)]
- Added support for Azure Active Directory Device Code Flow authentication [#597](https://github.com/dotnet/SqlClient/pull/597) [[Read more](#Azure-Active-Directory-Device-Code-Flow-authentication)]
- Added support for Azure Active Directory Managed Identity authentication [#730](https://github.com/dotnet/SqlClient/pull/730) [[Read more](#Azure-Active-Directory-Managed-Identity-authentication)]

> - Added new APIs to enable customizations for "Active Directory Interactive" authentication experience [#597](https://github.com/dotnet/SqlClient/pull/597) [Read more](#Azure-Active-Directory-Interactive-authentication-enhancements)

Reviewer: ```suggestion
- Added APIs to enable customizing the UI for "Active Directory Interactive" authentication [#597](https://github.com/dotnet/SqlClient/pull/597) [Read more](#Azure-Active-Directory-Interactive-authentication-enhancements)
```

- Added TraceLogging in Native SNI to extend `SqlClientEventSource` support [#650](https://github.com/dotnet/SqlClient/pull/650) [[Read more](#Trace-Logging-support-in-Native-SNI)]
- Updated Microsoft.Data.SqlClient.SNI (.NET Framework dependency) and Microsoft.Data.SqlClient.SNI.runtime (.NET Core/Standard dependency) version to v2.1.0 with trace logging implementation [#705](https://github.com/dotnet/SqlClient/pull/705)
- Added the "Command Timeout" connection string property to set a default timeout for all commands executed with the connection [#722](https://github.com/dotnet/SqlClient/pull/722) [[Read more](#Command-Timeout-connection-string-property)]
- Microsoft.Data.SqlClient symbols are now source-linked [#789](https://github.com/dotnet/SqlClient/pull/789) [[Read more](#Source-Linking-of-Microsoft.Data.SqlClient-symbols)]

> - Added new API to clear cached access tokens from token provider [#800](https://github.com/dotnet/SqlClient/pull/800)

Reviewer: ```suggestion
- Added an API to clear cached access tokens from the token provider [#800](https://github.com/dotnet/SqlClient/pull/800)
```

> - Added implementation of `SqlFacetAttribute` [#757](https://github.com/dotnet/SqlClient/pull/757)

Reviewer: ```suggestion
- Added `SqlFacetAttribute` implementation [#757](https://github.com/dotnet/SqlClient/pull/757)
```

### Fixed
- Fixed Enclave session cache issue with Azure Database [#686](https://github.com/dotnet/SqlClient/pull/686)
- Fixed pooled connection re-use on access token expiry issue when using Active Directory authentication modes [#635](https://github.com/dotnet/SqlClient/pull/635)
- Fixed transient fault handling for Pooled connections [#637](https://github.com/dotnet/SqlClient/pull/637)
- Fixed SPN generation issue when no port is provided [#629](https://github.com/dotnet/SqlClient/pull/629)
- Fixed missing null checks for `SqlErrors` in `SqlException` for .NET Framework implementation [#698](https://github.com/dotnet/SqlClient/pull/698)
- Fixed unobserved exception issue when a timeout occurs before a faulted task completes with an exception [#688](https://github.com/dotnet/SqlClient/pull/688) [#773](https://github.com/dotnet/SqlClient/pull/773)

> - Fixed an issue where driver continues to prompt for credentials when using Azure Active Directory authentication [#770](https://github.com/dotnet/SqlClient/pull/770)

Reviewer: ```suggestion
- Fixed an issue where the driver continues to prompt for credentials when using Azure Active Directory authentication and cached credentials should have been used [#770](https://github.com/dotnet/SqlClient/pull/770)
```

> - Fixed `InvalidOperationException` and `NotSupportedException` errors from driver due to `WriteAsync` collisions [#796](https://github.com/dotnet/SqlClient/pull/796)

Reviewer: ```suggestion
- Fixed `InvalidOperationException` and `NotSupportedException` errors due to `WriteAsync` collisions [#796](https://github.com/dotnet/SqlClient/pull/796)
```

> - Fixed bug in `ExecuteXmlReaderAsync` implementation [#782](https://github.com/dotnet/SqlClient/pull/782)

Reviewer: ```suggestion
- Fixed incorrect Settings.Async flag in `ExecuteXmlReaderAsync` [#782](https://github.com/dotnet/SqlClient/pull/782)
```

> - Fix a regression in Windows Integrated authentication when using managed networking [#777](https://github.com/dotnet/SqlClient/pull/777)

Reviewer: ```suggestion
- Fixed a regression in Windows Integrated authentication when using managed networking [#777](https://github.com/dotnet/SqlClient/pull/777)
```

- Fixed Bulk Copy Async deadlock issues with custom `IDataReader` when using `SqlDataReader` internally [#779](https://github.com/dotnet/SqlClient/pull/779)

> - Fix serialization issue with `SqlException` in .NET Core [#780](https://github.com/dotnet/SqlClient/pull/780)

Reviewer: ```suggestion
- Fixed a serialization issue with `SqlException` in .NET Core [#780](https://github.com/dotnet/SqlClient/pull/780)
```

### Changes

> - Performance improvements by fixing unnecessary allocations with EventSource implementation [#684](https://github.com/dotnet/SqlClient/pull/684)

Reviewer: ```suggestion
- Performance improvements by fixing unnecessary allocations in EventSource implementation [#684](https://github.com/dotnet/SqlClient/pull/684)
```

- Reverted changes to return empty `DataTable` from `GetSchemaTable` to return null as before. [#696](https://github.com/dotnet/SqlClient/pull/696)
- Removed multiple `CacheConnectionStringProperties` calls when setting `ConnectionString` properties [#683](https://github.com/dotnet/SqlClient/pull/683)

> - Code Improvements by only checking inexact match when no exact match is found for an embedded resource [#668](https://github.com/dotnet/SqlClient/pull/668)

Reviewer: ```suggestion
- Code improvements by checking for an inexact match only when no exact match is found for an embedded resource [#668](https://github.com/dotnet/SqlClient/pull/668)
```

- Updated versions of `Microsoft.IdentityModel` package dependencies [#794](https://github.com/dotnet/SqlClient/pull/794)

### Cross-Platform support for Always Encrypted v1
Microsoft.Data.SqlClient v2.1 extends support for Always Encrypted on the following platforms:

| Target Framework | Platform | New Feature support |
|------------------|----------|----|
| .NET Framework v4.6+ | Windows | *No changes* |
| .NET Core 2.1+ | All Operating Systems | **Yes, Support on Linux/MacOS** |
| .NET Standard 2.0+ | All Operating Systems | **Yes, Support on All Operating Systems** |

### Cross-Platform support for Always Encrypted with Secure Enclaves
Microsoft.Data.SqlClient v2.1 extends support for Always Encrypted (including with secure enclaves) on the following platforms:

| Target Framework | Platform | New Feature support |
|------------------|----------|----|
| .NET Framework v4.6+ | Windows | *No changes* |
| .NET Core 2.1+ | All Operating Systems | **Yes, Support on Linux/MacOS** |
| .NET Standard 2.1+ | All Operating Systems | **Yes, Support on All Operating Systems** |

> > .NET Standard 2.0 does not support Always Encrypted with Secure Enclaves.

Reviewer: ```suggestion
> Always Encrypted with secure enclaves is not supported on .NET Standard 2.0.
```

EventSource.SendCommand(eventSource, (EventCommand)16384, null);

// Enables both trace and flow events:
EventSource.SendCommand(eventSource, (EventCommand)(8192 | 16384), null);
```

### "Command Timeout" connection string property
Microsoft.Data.SqlClient v2.1 introduces the "Command Timeout" connection string property to override the default of 30 seconds. The timeout for individual commands can be overridden using the `CommandTimeout` property on the SqlCommand.

Connection string examples:

`"Server:{serverURL}; Initial Catalog={db}; Integrated Security=true; Command Timeout=60"`

### Removal of symbols from Native SNI
With Microsoft.Data.SqlClient v2.1, we've removed the symbols introduced in [v2.0.0](https://www.nuget.org/packages/Microsoft.Data.SqlClient.SNI/2.0.0) from [Microsoft.Data.SqlClient.SNI.runtime](https://www.nuget.org/packages/Microsoft.Data.SqlClient.SNI.runtime) NuGet starting with [v2.1.1](https://www.nuget.org/packages/Microsoft.Data.SqlClient.SNI.runtime/2.1.1). The public symbols are now published to Microsoft Symbols Server for tools like BinSkim that require access to public symbols.

### Source-Linking of Microsoft.Data.SqlClient symbols

> Starting with Microsoft.Data.SqlClient v2.1, Microsoft.Data.SqlClient symbols are source-linked and published to Microsoft Symbols Server for enhanced debugging experience without the need to download source code.

Reviewer: ```suggestion
Starting with Microsoft.Data.SqlClient v2.1, Microsoft.Data.SqlClient symbols are source-linked and published to the Microsoft Symbols Server for an enhanced debugging experience without the need to download source code.
```
